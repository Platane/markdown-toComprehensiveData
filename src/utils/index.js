import type { AbstractTree } from '../type'

/**
 * transform to human readable tree
 */
export const toString = (tree: AbstractTree, prefix?: string = '') => {
    return [
        prefix +
            tree.type +
            ' ' +
            (tree.text ? ` '${tree.text}'` : '') +
            (tree.src || ''),
        ...(tree.children && tree.children.length
            ? [
                  //   prefix + '[',
                  ...[].concat(
                      ...tree.children.map(c => toString(c, prefix + '  '))
                  ),
                  //   prefix + ']',
              ]
            : []),
    ].join('\n')
}

/**
 * return the first subtree which validate the matching function
 */
export const find = <T: AbstractTree>(
    match: (tree: T, path: string[], ancestors: T[]) => boolean,
    tree: T,
    ancestors?: T[] = []
): T | null =>
    match(tree, [tree, ...ancestors].reverse().map(x => x.type), ancestors)
        ? tree
        : tree.children.reduce(
              (f, c) => f || find(match, c, [tree, ...ancestors]),
              null
          )

/**
* return all the subtree which validate the matching function
*/
export const findAll = <T: AbstractTree>(
    match: (tree: T, path: string[], ancestors: T[]) => boolean,
    tree: T,
    matched?: T[] = []
): T[] => {
    const f = find(
        (x, ...rest) => match(x, ...rest) && matched.indexOf(x) === -1,
        tree
    )
    return f ? findAll(match, tree, [...matched, f]) : matched
}

/**
* remove the subtree
*/
export const prune = <T: AbstractTree>(tree: T, subTreeToRemove: T): T => ({
    ...tree,
    children: tree.children
        .filter(x => x != subTreeToRemove)
        .map(x => prune(x, subTreeToRemove)),
})

/**
* read the tree and concat the text node
*/
export const extractText = (tree): string => {
    let next = tree.children.reduce((s, c) => s + extractText(c), '')

    if ('\n' === next[next.length - 1]) next = next.slice(0, -1)

    switch (tree.type) {
        case 'text':
            return tree.text || ''

        case 'textBlock':
            return next + '\n'

        case 'softbreak':
            return '\n'

        default:
            return next
    }
}
