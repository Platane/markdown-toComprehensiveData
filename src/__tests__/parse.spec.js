import { parse } from '../index'
import { toString } from '../utils'
import fs from 'fs'
import path from 'path'

fs
    .readdirSync(path.join(__dirname, './samples'))
    .forEach(filename =>
        it(`should parse the file "${filename}"`, () =>
            expect(
                toString(
                    parse(
                        fs
                            .readFileSync(
                                path.join(__dirname, './samples', filename)
                            )
                            .toString()
                    )
                )
            ).toMatchSnapshot())
    )
