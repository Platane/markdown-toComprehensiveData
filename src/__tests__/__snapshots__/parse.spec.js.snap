// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse the file "13k.md" 1`] = `
"doc 
  textBlock 
    link 
      image https://platane.github.io/js13k-2017/game.gif
  textBlock 
    italic 
      text  'My entry for the 13k js 2017'
  textBlock 
    link 
      text  'play!'
  textBlock 
    text  'or go to the '
    link 
      text  'js13k entry page'
  heading 
    text  'Gameplay'
  textBlock 
    text  'You are lost in a museum.'
  textBlock 
    text  'Navigate from a painting to another in order to find your way out.'
  heading 
    text  'Controls'
  list 
    textBlock 
      text  'desktop : use the mouse + space to move forward'
    textBlock 
      text  'mobile : use the gyroscope + touch the screen to move forward'
    textBlock 
      text  'cardboard : use the gyroscope + magnetic button to move forward'
  heading 
    text  'Inspiration'
  textBlock 
    text  'I challenge myself to put as much image as possible in the game, in form of widly known piece of art.'
  textBlock 
    text  'To do so, I used a genetic algorithm to describe an arbitrary image as a set of colored dot.'"
`;

exports[`should parse the file "big.md" 1`] = `
"doc 
  textBlock 
    text  'This is my first paragraph.'
    softbreak 
    text  'This is a second line'
  textBlock 
    text  'Second paragraph with '
    italic 
      text  'italic'
    text  ' and '
    bold 
      text  'bold'
    text  ' stuffs.'
  imageGroup 
    image 'myImage1.png'
    image 'myImage2.png'
  textBlock 
    text  'Third paragraph '
    link 
      text  'click me'
  textBlock 
    link 
      image 'myImage3.png'
    softbreak 
    link 
      image 'myImage4.png'"
`;

exports[`should parse the file "bockquote.md" 1`] = `
"doc 
  codeBlock  'This is code
'
  quoteBlock 
    textBlock 
      text  'yolo'
      softbreak 
      code  'another quote'
    quoteBlock 
      textBlock 
        text  'asds'
  textBlock 
    text  'and this stuff'"
`;

exports[`should parse the file "ex.md" 1`] = `
"doc 
  textBlock 
    text  'This is my first paragraph.'
    softbreak 
    text  'This is a second '
    bold 
      text  'line'
    text 
  imageGroup 
    image link/to/image
    image link/to/another/image"
`;

exports[`should parse the file "imageBlocks.md" 1`] = `
"doc 
  imageGroup 
    image 'myImage1.png'
  imageGroup 
    image 'myImage1.png'
    image 'myImage2.png'
  textBlock 
    link 
      image 'myImage3.png'
    softbreak 
    link 
      image 'myImage4.png'"
`;

exports[`should parse the file "italicBold.md" 1`] = `
"doc 
  textBlock 
    text  'normal and '
    italic 
      text  'italic'
    text  ' and '
    bold 
      text  'bold'
    text  ' and '
    italic 
      text 
      bold 
        text  'bold and italic'
      text "
`;

exports[`should parse the file "list.md" 1`] = `
"doc 
  list 
    textBlock 
      text  '1'
    textBlock 
      text  '2'
    textBlock 
      text  '3'
  list 
    textBlock 
      text  '1'
    textBlock 
      text  '2'
    textBlock 
      text  '3 asd asdssad'
      softbreak 
      text  'asdasdasd'
    textBlock 
      text  '4'"
`;

exports[`should parse the file "meta.md" 1`] = `
"doc 
  hr 
  heading 
    text  'sadaspas'
    softbreak 
    text  'sapasd: asdas'
  textBlock 
    text  'asdasd'"
`;

exports[`should parse the file "meta2.md" 1`] = `
"doc 
  hr 
  heading 
    text  'sadaspas'
    softbreak 
    text  'sapasd: asdas:123213'
  textBlock 
    text  'asdasd'"
`;

exports[`should parse the file "title.md" 1`] = `
"doc 
  heading 
    text  'one'
  heading 
    text  'two'
  heading 
    text  'three'
  heading 
    text  'four'"
`;
